### ISSUE TITLE
recruitment deviation objects should be Nyrs-1?

### ISSUE THREAD

**Issue by gavinfay** [thefaylab/hydra_sim#27]
Recruitment in the first year is confounded with the initial N in the first size bin.
@sgaichas I addressed this (for now) by reducing the length of the recdev objects, thus recruitment only happens in years 2:Nyr.
Does this make sense?
To get things to work properly, the relevant `.pin` file vectors trimming by 1 year for each species if we stay with this. (done manually for now)

**Comment by sgaichas** [comment]
This sounds reasonable.
I may be overthinking it, but does this assume N@sizebin1=Nrecruits? Recruits only go into bin1 but not all of bin1 is recruits since there can in theory be fish in that bin prior to recruitment (from a previous year), depending on growth combined with the bin size. 
With Nstepsyr = 1, only the fastest growing species is out of the first bin in the first timestep, the others will have some proportion still in that bin in the next timestep (which is the next year) so recruits are added to a nonzero bin1. 
Would an alternative be to define initial N at sizebin1 as "pre-recruitment" ? Then recruits can still get added in year 1. I'm not sure how to split that hair though, or if it makes anything easier.

**Comment by gavinfay** [comment]
Yes, length bin 1 contains the new recruits plus those left from last year. But, as recruitment is the first thing that happens, in the initial year there is no way to distinguish the new recruits from those prior to the model start. As the N in length bin 1 is an estimated parameter then these are already being accounted for. One could use the expected recruitment (e.g. average for option 9) added on to N1, but can't have this first year value be an independent estimated parameter (rec dev).

**Comment by sgaichas** [comment]
Sounds good. The new pin file code makes the recdev vector for Nyrs-1 https://github.com/thefaylab/hydradata/commit/6861e68626677a785b4dfe91e01a5dc1c91f16a5
and new input files and data object (that also fixed inpN and the indicatorFisheryq order) reflect this change in 265092a9196f52dc0792b2aca98ce4e1b09f9444

**Comment by sgaichas** [comment]
I'm not sure whether this change has unexpectedly altered models with more than one timestep per year--NOBA Cod fit ok in the previous model but is having problems now.

But other species don't share this and cod has its own fishery so was there a change in fishery specification as well? results from 5 bin model below

*prior to change* (should be identical inputs): 
![image](https://user-images.githubusercontent.com/8172302/194592220-6d4b9e9f-ffd1-48ff-9795-571a58c0d7ed.png)

*current recruitment config*:
![image](https://user-images.githubusercontent.com/8172302/194592308-8cea1b7c-505a-4e16-aaa0-720bf6e19c41.png)

mortality, current config (no equivalent plot from earlier)
![image](https://user-images.githubusercontent.com/8172302/194593027-69c29b18-606e-4a93-ae3b-8b065f73b4e5.png)

![image](https://user-images.githubusercontent.com/8172302/194593064-6a74450d-6abb-45ef-b572-583834d5a21b.png)

Fishery selectivity is flat for the 5 bin model
![image](https://user-images.githubusercontent.com/8172302/194593784-20efe1b8-068d-4111-a785-71b377426992.png)


The 10 bin model in the current code shows a much less severe pattern (but still worse than before for cod) and estimates fishery selectivity that looks more reasonable

![image](https://user-images.githubusercontent.com/8172302/194593929-ac149c02-176d-46c9-84f3-4096e31abae1.png)
![image](https://user-images.githubusercontent.com/8172302/194593976-51d7a5cb-354d-4362-8860-f53262cd4910.png)
![image](https://user-images.githubusercontent.com/8172302/194594007-e3689aff-38d6-4a45-a5a5-963cf77cc9d3.png)

to be fair, actual selectivity for cod in NOBA is strange so maybe this new configuration is more accurately finding it difficult to deal with...

**Comment by sgaichas** [comment]
The 5 bin model seems more sensitive to the M1 splitting calculation. 
In my local code I reverted to 
```
 M1(area, spp)  = 1.0 - pow((1.0 - M1ann(area, spp)), (1.0 / Nstepsyr))
```
resulting in M1 = 0.2085 rather than 0.2

With no other changes; the cod problem is less severe. It is behaving more like the 10 bin model (but still worse fit for cod).

![image](https://user-images.githubusercontent.com/8172302/194714623-431a0f17-2a49-459e-9c65-2abb8913989a.png)
![image](https://user-images.githubusercontent.com/8172302/194714816-f4763e76-1c94-46c1-b7c9-d2f263414811.png)
![image](https://user-images.githubusercontent.com/8172302/194714904-71ce2369-5c54-4590-9383-9410bad4dabf.png)


skill plot for B shows it is picking up signals for haddock and herring pretty well
![image](https://user-images.githubusercontent.com/8172302/194714725-6ca8bcac-b59e-49c7-a84e-56f7291df93c.png)

**Comment by gavinfay** [comment]
I am having some similar issues with regard to recruitment and selectivity for some of the GB stocks with the 5 bin model. I haven't actually run the 10bin version of that with the recent files yet.
I will revert to your implementation of M1. I also have a local change where M1 is an (estimable) parameter, but fixing it for now. (so identical to the data file implementation). Though maybe it would also make sense to estimate something related to prey vulnerability if we were going to start doing this to truly allow the interactions to be estimated during fitting, e.g. like Vanessa's model. A discussion point for next week for sure.

**Comment by gavinfay** [comment]
It probably also makes sense to have a 'last yr' specification for recruitment devs so that those at the end of the time series aren't estimated if they represent year classes not yet observed fully in fishery/surveys. (e.g. SS)

**Comment by sgaichas** [comment]
I haven't made a 10 bin GB dataset yet, we would have to make some decisions about the input stomach weights and a few other currently fixed input parameters to make that work. The 10 bin NOBA takes many hours to run on my computer so may not be worth your time right now!
Yes that sounds good on the last year.

**Comment by gavinfay** [comment]
Ah. Clearly my brain is mush. Sorry!

**Comment by sgaichas** [comment]
My brain was too mushy to make those decisions for 10 bin.  :-) still is!
The newest 5 bin GB data inputs do correct the fishery catch at length and have lower inpN than 100 though, so if you haven't tried them they may change things. Sorry about late changes to data...
